src/ShallowWaterExplicit.jl

```diff
@@ -0,0 +1,171 @@
+  # RTMM       : H(div) mass matrix, ∫β⋅βdΩ, ∀β∈ H(div,Ω)
+  # L2MM       : L² mass matrix, ∫γγdΩ, ∀γ∈ L²(Ω)
+  # dt         : time step
+  # leap_frog  : do leap frog time integration for the first step (boolean)
+  # dΩ         : measure of the elements
+
+  n = get_normal_vector(model)
+  # explicit step for provisional velocity, uₚ
+  dt1 = dt
+  if leap_frog
+    dt1 = 2.0*dt
+  end
+
+  # 1.1: the mass flux
+  b₁(v)  = ∫(v⋅u₁*h₁)*dΩ
+  rhs1   = assemble_vector(b₁, V)
+  copy!(get_free_dof_values(F), rhs1)
+  ldiv!(RTMMchol, get_free_dof_values(F))
+  # 1.2: the bernoulli function
+  b₂(q)  = ∫(q*(0.5*u₁⋅u₁ + g*h₁))*dΩ
+  rhs2   = assemble_vector(b₂, Q)
+  copy!(get_free_dof_values(ϕ), rhs2)
+  ldiv!(L2MMchol, get_free_dof_values(ϕ))
+  # 1.3: the potential vorticity
+  a₁(r,s) = ∫(s*h₁*r)dΩ
+  c₁(s)   = ∫(perp(n,∇(s))⋅(u₁) + s*f)dΩ
+  H1h     = assemble_matrix(a₁, R, S)
+  rhs_q₁  = assemble_vector(c₁, S)
+  op      = AffineFEOperator(R, S, H1h, rhs_q₁)
+  q₁      = solve(op)
+  # 1.4: solve for the provisional velocity
+  b₃(v)  = ∫(v⋅uₘ - dt1*(q₁ - τ*u₁⋅∇(q₁))*(v⋅⟂(F,n)))dΩ + ∫(dt1*DIV(v)*ϕ)*dω
+  rhs3   = assemble_vector(b₃, V)
+  copy!(get_free_dof_values(uₚ), rhs3)
+  ldiv!(RTMMchol, get_free_dof_values(uₚ))
+  # 1.5: solve for the provisional depth
+  b₄(q)  = ∫(q*hₘ)dΩ - ∫(dt1*q*DIV(F))*dω
+  rhs4   = assemble_vector(b₄, Q)
+  copy!(get_free_dof_values(hₚ), rhs4)
+  ldiv!(L2MMchol, get_free_dof_values(hₚ))
+
+  # 2.1: the mass flux
+  b₅(v)  = ∫(v⋅u₁*(2.0*h₁ + hₚ)/6.0 + v⋅uₚ*(h₁ + 2.0*hₚ)/6.0)*dΩ
+  rhs5   = assemble_vector(b₅, V)
+  copy!(get_free_dof_values(F), rhs5)
+  ldiv!(RTMMchol, get_free_dof_values(F))
+  # 2.2: the bernoulli function
+  b₆(q)  = ∫(q*((u₁⋅u₁ + u₁⋅uₚ + uₚ⋅uₚ)/6.0 + 0.5*g*(h₁ + hₚ)))*dΩ
+  rhs6   = assemble_vector(b₆, Q)
+  copy!(get_free_dof_values(ϕ), rhs6)
+  ldiv!(L2MMchol, get_free_dof_values(ϕ))
+  # 2.3: the potential vorticity
+  a₂(r,s) = ∫(s*hₚ*r)dΩ
+  c₂(s)   = ∫(perp(n,∇(s))⋅(uₚ) + s*f)dΩ
+  H2h     = assemble_matrix(a₂, R, S)
+  rhs_q₂  = assemble_vector(c₂, S)
+  op      = AffineFEOperator(R, S, H2h, rhs_q₂)
+  q₂      = solve(op)
+  # 2.4: solve for the final velocity
+  b₇(v)  = ∫(v⋅u₁ - 0.5*dt*(q₁ - τ*u₁⋅∇(q₁) + q₂ - τ*uₚ⋅∇(q₂))*(v⋅⟂(F,n)))dΩ + ∫(dt*DIV(v)*ϕ)*dω
+  rhs7   = assemble_vector(b₇, V)
+  copy!(get_free_dof_values(u₂), rhs7)
+  ldiv!(RTMMchol, get_free_dof_values(u₂))
+  # 2.5: solve for the final depth
+  b₈(q)  = ∫(q*h₁)dΩ - ∫(dt*q*DIV(F))*dω
+  rhs8   = assemble_vector(b₈, Q)
+  copy!(get_free_dof_values(h₂), rhs8)
+  ldiv!(L2MMchol, get_free_dof_values(h₂))
+end
+
+function shallow_water_time_stepper(model, order, degree, h₀, u₀, f₀, g, nstep, diag_freq, dump_freq, dt, τ)
+  # Forward integration of the shallow water equations using a supplied method
+  Ω = Triangulation(model)
+  dΩ = Measure(Ω, degree)
+  dω = Measure(Ω, degree, ReferenceDomain())
+  quad_cell_point = get_cell_points(dΩ.quad)
```

Comment:
I would say these four lines are no longer needed right ?

Reply:
I need the top three to pass down to the explicit solver, but not the others below, have removed...